### PR TITLE
Rework functionality of leveled nodes

### PR DESCRIPTION
**Pull request by Wuzzy2**
## The problem

Falling nodes in `builtin` currently include code that is apparently supposed to handle a leveled node falling into another leveled node of the same type by increasing the node level.

However, this code is completely broken as it made several nonsensical assumptions about the Lua API: E.g. there is `node.level` (which doesn't exist) or `node_definition.leveled` is used to check whether a node is leveled (must check `paramtype2` instead).

Furthermore, the function `minetest.add_node_level`, on which the falling node code relies on, it also broken. If you add a high level value to a node with an already high level, there will be a wrap-around. E.g. `minetest.add_node_level(pos, 50)` when the node level was already 120 results in a wrap-around.

## What this PR does

* When a falling leveled node falls into another leveled node of the same type, the destination node's level is increased (which the code *already* tried to do, but failed to)
* Fixes `minetest.add_node_level` wrapping around. This is done by adding hard caps at 0 and `LEVELED_MAX` (=127).

## Limitations and oddities

1) A node can grow higher than 1 node height with this, because leveled nodes can grow up to 127, which is almost the double height of a single node. But this creates all sorts of oddities and weirdnesses in the falling behavior which are probably not easy to fix.
2) If a leveled node of height 0 falls into a leveled node with heigh 0, the resulting node will not grow. This is intended.
3) While a leveled node falls, the rendering will always use the default height of the node (it pretends `param2` is always 0). If `param2` is not 0, the rendered height of the falling leveled node might be incorrect. This is a separate bug that is out of scope for this PR.

Question: Should I allow nodes to have their level grow larger than 64 (=1 node height) with falling nodes? It might be a good idea to cap the node height to 64 with using falling nodes alone, to keep the code simple.

## HOW TO TEST

* Grab `devtest`
* Take some nodes of type `testnodes:nodebox_leveled`
* Place several of these around
* Set the node height with the Param2 Tool
* Make them fall into each other with the Falling Node Tool
* Try out different `param2` values

### CODE REVIEW

File: src/mapnode.cpp [hunk 1]
```diff
@@ -642,9 +642,19 @@ u8 MapNode::setLevel(const NodeDefManager *nodemgr, s8 level)
...
-	level += add;
-	return setLevel(nodemgr, level);
+	s16 tmpLevel = getLevel(nodemgr);
+	tmpLevel += add;
+	u8 rest = 0;
+	if (tmpLevel > LEVELED_MAX) {
+		rest = tmpLevel - LEVELED_MAX;
+		tmpLevel = LEVELED_MAX;
+	} else if (tmpLevel < 0) {
+		rest = -tmpLevel;
```
Comment (sfan5) [inline src/mapnode.cpp:652]: the function can only return unsigned numbers so this doesn't make sense

**Comment by Wuzzy2** [comment]
I reworked `addLevel` and `setLevel` as requested.

Should I add `leveled_max`? Or is this better left for a later PR?

**Comment by paramat** [comment]
> It might be a good idea to cap the node height to 64

I agree. I suspect resulting nodes being higher than 1 node will cause complexities. Strictly accurate addition of level is not important or necessary.

**Comment by Wuzzy2** [comment]
We would have to keep the current max. level of 127, however, because of backwards-compability.

**Comment by sfan5** [comment]
> Should I add `leveled_max`? Or is this better left for a later PR?

Yes, should be one PR to completely fix leveled node behavior IMO.

**Comment by Wuzzy2** [comment]
`leveled_max` added.

File: src/mapnode.cpp [hunk 2]
```diff
@@ -603,8 +603,8 @@ u8 MapNode::getLevel(const NodeDefManager *nodemgr) const
 		if (level)
 			return level;
 	}
-	if (f.leveled > LEVELED_MAX)
-		return LEVELED_MAX;
+	if (f.leveled > f.leveled_max)
+		return f.leveled_max;
```
Comment (sfan5) [inline src/mapnode.cpp:607]: Please add `// Return static value from nodedef if param2 isn't used for level`
This confused me at first

You can also replace the if + returns with `return MYMIN(f.leveled, f.leveled_max);`

File: builtin/game/falling.lua [hunk 3]
```diff
@@ -196,10 +196,10 @@ core.register_entity(":__builtin:falling_node", {
 	try_place = function(self, bcp, bcn)
 		local bcd = core.registered_nodes[bcn.name]
 		-- Add levels if dropped on same leveled node
-		if bcd and bcd.leveled and
+		if bcd and bcd.paramtype2 == "leveled" and
 				bcn.name == self.node.name then
-			local addlevel = self.node.level
-			if not addlevel or addlevel <= 0 then
+			local addlevel = self.node.param2 % 128
+			if bcd.leveled and addlevel == 0 then
```
Comment (sfan5) [inline builtin/game/falling.lua:202]:
```suggestion
			if addlevel == 0 and bcd.leveled then
```

File: src/nodedef.cpp [hunk 4]
```diff
@@ -435,6 +436,7 @@ void ContentFeatures::serialize(std::ostream &os, u16 protocol_version) const
 		writeU16(os, connects_to_id);
 	writeARGB8(os, post_effect_color);
 	writeU8(os, leveled);
+	writeU8(os, leveled_max);
```
Comment (sfan5) [inline src/nodedef.cpp:439]: this is not backwards compatible

File: doc/lua_api.txt [hunk 5]
```diff
@@ -7002,7 +7002,11 @@ Used by `minetest.register_node`.
         -- Only valid for "nodebox" drawtype with 'type = "leveled"'.
```
Comment (sfan5) [inline doc/lua_api.txt:7002]: please change the definition of leveled above here to `leveled = 0` (the default value)

**Comment by Wuzzy2** [comment]
> this is not backwards compatible

How to make this backwards-compatible then?

I don't see how I could add a new field to the nodedef without touching the data.

**Comment by sfan5** [comment]
> How to make this backwards-compatible then?

It's not backwards-compatible because it's right in the middle of serialized data.
You have to add it at the end and also include code to handle the case where the value is not present

File: src/nodedef.cpp [hunk 6]
```diff
@@ -586,6 +588,7 @@ void ContentFeatures::deSerialize(std::istream &is)
 
 	try {
 		node_dig_prediction = deSerializeString(is);
+		leveled_max = readU8(is);
```
Comment (sfan5) [inline src/nodedef.cpp:591]: `readU8` unfortunately doesn't throw an exception so this doesn't work
The following should work (not sure if function name is correct):
```c++
    u8 tmp = readU8(is);
    if (is.is_eof())
        throw SerializationError();
    leveled_max = tmp;
```

**Comment by Wuzzy2** [comment]
Needs testing now. I'm not sure it's without bugs.

File: src/nodedef.cpp [hunk 7]
```diff
@@ -587,6 +589,11 @@ void ContentFeatures::deSerialize(std::istream &is)
 	try {
 		node_dig_prediction = deSerializeString(is);
 	} catch(SerializationError &e) {};
+
+	u8 tmp_leveled_max = readU8(is);
+	if (is.eof())
+		throw SerializationError("Nodedef: Missing leveled_max");
```
Comment (sfan5) [inline src/nodedef.cpp:595]: this whole chunk was supposed to be in the try {} obviously

**Comment by Wuzzy2** [comment]
I'm obviously too incompetent to do this.

File: src/nodedef.cpp [hunk 8]
```diff
@@ -586,6 +588,10 @@ void ContentFeatures::deSerialize(std::istream &is)
 
 	try {
 		node_dig_prediction = deSerializeString(is);
+		u8 tmp_leveled_max = readU8(is);
+		if (is.eof()) /* readU8 doesn't throw exceptions so we have to do this */
```
Comment (SmallJoker) [inline src/nodedef.cpp:592]: What if `readU8` reads the last byte? Shouldn't this be checked before reading the byte?

Comment (sfan5) [reply src/nodedef.cpp:592]: `eof` is only set after readU8 unsuccessfully tried to read the last byte.
At least that's the behaviour I remember, might be wrong.

Comment (SmallJoker) [reply src/nodedef.cpp:592]: Indeed, you remembered correctly.
Logging before and after `readU8`:
```
before? 0
after? 1
```

**Comment by sfan5** [comment]
Nodedef used for testing (tested with/without `buildable_to`, with/without `leveled_max`):
```lua
local snowbox = {
	type = "leveled",
	fixed = {-0.5, -0.5, -0.5, 0.5, -0.5, 0.5},
}
minetest.register_node("test2:snow", {
	description = "Test Snow",
	tiles = {"default_snow.png"},
	inventory_image = "default_snowball.png",
	wield_image = "default_snowball.png",
	paramtype = "light",
	paramtype2 = "leveled",
	buildable_to = true,
	floodable = true,
	walkable = true,
	drawtype = "nodebox",
	leveled = 16,
	leveled_max = 64,
	node_box = snowbox,
	collision_box = snowbox,
	selection_box = snowbox,
	groups = {crumbly = 3, falling_node = 1, snowy = 1},
	node_placement_prediction = "",
	on_place = function(itemstack, placer, pt)
		local pos = pt.under
		local node = minetest.get_node(pos)
		if node.name ~= "test2:snow" then return minetest.item_place(itemstack, placer, pt) end
		minetest.add_node_level(pos, 16)
		return itemstack
	end,
})
```

File: builtin/game/falling.lua [hunk 9]
```diff
@@ -122,6 +123,24 @@ core.register_entity(":__builtin:falling_node", {
...
+		local nb_types = {fixed=true, leveled=true, connected=true}
+		if def.drawtype == "nodebox" and def.node_box and
+			nb_types[def.node_box.type] then
+			local box = table.copy(def.node_box.fixed)
+			if type(box[1]) == "table" then
+				box = #box == 1 and box[1] or nil -- We can only use a single box
+			end
+			if box then
+				if def.paramtype2 == "leveled" and (self.node.level or 0) > 0 then
+					box[5] = -0.5 + self.node.level / 64
```
Comment (SmallJoker) [inline builtin/game/falling.lua:136]: Shouldn't this now be `self.node.level / def.leveled_max` plus another check above?

Comment (sfan5) [reply builtin/game/falling.lua:136]: https://github.com/minetest/minetest/blob/7d3972a5049324f776ab008894c34569641f0073/src/mapnode.cpp#L248-L249

**Comment by SmallJoker** [comment]
Fixes #8307

**Pull request merged**